### ISSUE TITLE
Syntax to use multiple objects at once

### ISSUE THREAD

**Issue by LordYuuma** [spk121/guile-gi#12]
With `with-object` (alias `using`) we already have a quite useful syntax to set multiple properties, add, remove and (unblock) multiple signals and invoke multiple methods **of one object**. However, at times the exact opposite is needed, binding a single signal of many objects or even interacting with arbitrary objects in an arbitrary manner.

It seems appropriate – since we would otherwise use `let` or `let*` to bind such objects – to construct `let-objects` and `let-objects*` on top of them, using a replacement similar to that present in `with-object`. Some questions should be asked first, though.

1. How should we identify the objects being used? Pass them as extra parameter? If so, how do we represent property getters, which in with-object are a single identifier? Merge the object and the identifier, e.g. `obj::some-method`?
2. Should the replacements be recursive? If so, how do we interact with variables from the outside? If not, how do we capture the objects on the inside?
3. Should we add extra syntax to `let-objects` (perhaps `with-object` as well) for methods that deal with multiple objects, such as `bind-property`?

**Comment by spk121** [comment]
Could this be comparable to `map` and `for-each`?

```
(map-object (method args) list-of-objects)  => list of return values
(map-object (connect! signal args) list-of-objects)
(for-each-object (method args) list-of-objects) => no return values
```

**Comment by LordYuuma** [comment]
Not quite. `with-object` is a sequence of functions applied to a single object. `map-object` and `foreach-object` are single functions applied to multiple objects and can already be expressed quite well with `map`/`foreach` and lambdas or `cut`/`cute`. What I want is something like the following:

    (let-objects ((window (create ...))
                  (button (create ...)))
      (add window button)
      (connect! button (clicked ...))
      (show window))

**Comment by spk121** [comment]
As you've written it, the lifetimes of the scheme objects are nominally the scope of the `let-objects`, and they should be garbage collected as they go out of scope.  The GObjects themselves can, of course, live on if the methods called appropriately increase their refcount or hook them into containers that exist from outside the scope of the `let-objects`.   And the return values, or perhaps all the return values except for the last call, are discarded.  Is that right?

If I understand it correctly, I think the common usage of such a form would then be to prep some GObjects and then use GObject methods to attach them to external objects.  Which is okay, but, it lacks the feel of a tail-call-friendly syntax.

Maybe, in your example, the constructed objects should be returned to the caller in a tail position?
```
(let-objects ((window (create ...))
              (button (create ...)))
  (add window button)
  (connect! button clicked (lambda (button) ...))
  (show window)
  (values window button) ; ?
)
```

**Comment by LordYuuma** [comment]
`let-objects` would work like `let`, so the scheme objects are valid for the scope of `let-objects` and not without it and the return value should be the value of the last expression. So you're right there.

That would be one use-case. Another would be in handlers, which interact with multiple objects, e.g. a switch causes one item do disappear and another one to appear or something like that. Here the GObjects wouldn't be created as in this example, but they'd be temprorarily referenced (or whatever we do with them on the Scheme side).

You're touching on a weakness, that `with-object` already has, and that is that we're not able to return the object itself. Maybe we should add a `this` block (or `%this` in case someone decides to have a "this" property in one of their libraries)? For `let-objects` it would then be `(%this window)` or `window::%this` and multiple objects could be `(%these window button other-button)`. Either way, I don't think this example really highlights the issue, because windows will be created and registered with a GtkApplication from which you can fetch them, so we often don't care about the window itself after the `show`. However, in cases where we do want the objects to be returned.

On a related note, I kinda wonder, if should make method/property/signal lookup possible in Scheme. Then we could do something like `(if (has-method? obj "method") (call-method obj "method" arg ...) (method obj arg ...))` and would save a lot of hacks.

**Comment by spk121** [comment]
A `has-method?` would be an easy addition, and likely useful.  `gir_method_lookup_full` does have some functionality to choose between looking up if a method belongs to this (child) class, or if it belongs to this class and all parent classes, though I think it is neither correct nor tested.

Is it probably three procedures: `has-method?` `has-property?` `has-signal?`

Letting `with-object` return the object might be useful.I like the idea of a `%this` block of some description. There are a couple of easy options
```
(with-object obj (method ...) %this) ; thus no property can be named %this
or
(with-object obj (method ...) (%this)) ; no method can be named %this
or
(with-object obj (method ...) obj)  ;  no property can be named obj
```
For `let-objects` I think `values` could suffice for returning objects.
For `let-objects`, you need to decide if it is `let` or `let*` under the hood.  It is probably `let*` otherwise you end up with some questions.  Can the second argument assume the existence of the first argument? Can the second argument have the same name as the first argument? And what about these perverse cases?

```
(let-objects ((alpha (create ...)
        (bravo))
or
(let-objects ((alpha (let-objects ((bravo (create ...)) bravo))
```

The disadvantage of the `let` in the name `let-objects` is that users might have high expectations of what it could do.  One might assume that a `let-objects` can contain a `let` or `let-objects`.  Maybe a `with-objects` syntax promises less.

**Comment by LordYuuma** [comment]
I think that checking whether a method exists somewhere in the hierarchy at all would suffice and we already rely on that being somewhat correct in `call-method`.

Yes, something like that. Then we could expand `(blob obj arg ...)` to

    (cond
      ((has-method? obj "blob") (call-method obj "blob" arg ...))
      ((has-property? obj "blob") (gobject-get-property obj "blob"))
      ((has-signal? obj "blob") (fire-signal #| what is this called? |# obj "blob"))
      (else (blob obj arg ...)))

assuming we want `let-objects` to be as generic as possible. That way, we could also use Scheme stuff like `values`, nestes `let`s and whatever. On the other hand, efficiency would take a huge dump. (Perhaps a less big one if we used the type instead of the object directly?)

Alternatively, we could introduce a `(with-generics methods properties signals body ...)` syntax (analogous to guile-gnome's `with-accessors` for properties), but then we would compromise on the scoping part of the let and if could lead to weird conflicts when a method of one class conflicts with the property or signal of another. It really is much easier if we only have a single type and a single object to take care of.

I think property `%this` is the best way to go. Properties and methods shouldn't have special symbols other than the dash to begin with, so the name should not be a problem.

To make things as easy as possible conceptually, `let-objects` should be based on `let` and a starred alternative based on `let*` should be given (as indicated by in the original post). I think the goal should be to translate the "header" to a simple `let(*)` expression and have that take care of duplicate bindings. That being said, I am aware that doing such a thing is not as easy as one might guess with `syntax-case` and I don't think `syntax-rules` will suffice.

`with-objects` and `with-objects*` would promise less, but regardless of the name, we have to solve these problems, perhaps without the inclusion of arbitrary scheme code. In the end, however, I want as little syntactic overhad as possible for accessors, methods and the like, not caring too much about what happens under the hood. And I think for that something like `let` is better than whatever `with-` construct we're coming up with. If only we had generics...

**Comment by LordYuuma** [comment]
Well, now that we have GOOPS, we can just use normal `let` et al.
